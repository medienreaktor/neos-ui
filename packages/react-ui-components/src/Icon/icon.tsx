import {IconName, IconPrefix, IconProp} from '@fortawesome/fontawesome-svg-core';
import {FontAwesomeIcon} from '@fortawesome/react-fontawesome';
import mergeClassNames from 'classnames';
import React, {PureComponent} from 'react';

import {PickDefaultProps} from '../../types';
import mapper from './mapper';

type IconPadding = 'none' | 'left' | 'right';
type IconColor = 'default' | 'warn' | 'error' | 'primaryBlue';

interface IconTheme {
    readonly icon: string;
    readonly 'icon--big': string;
    readonly 'icon--small': string;
    readonly 'icon--tiny': string;
    readonly 'icon--paddedLeft': string;
    readonly 'icon--paddedRight': string;
    readonly 'icon--spin': string;
    readonly 'icon--color-warn': string;
    readonly 'icon--color-error': string;
    readonly 'icon--color-primaryBlue': string;
}

export interface IconProps {
    /**
     * We use the react component FortAwesome provides to render icons.
     * we will pass down all props to the component via {...rest} to expose it's api
     * https://github.com/FortAwesome/react-fontawesome/blob/master/README.md
     */

    /**
     * The ID of the icon to render.
     */
    readonly icon: string;

    /**
     * The (accessibility) label for this icon
     */
    readonly label?: string;

    /**
     * Controls the padding around the icon in a standardized way.
     */
    readonly padded?: IconPadding;

    /**
     * An optional `className` to attach to the wrapper.
     */
    readonly className?: string;

    /**
     *  Adjust the color of the icon
     */
    readonly color?: IconColor;

    /**
     *  An optional css theme to be injected.
     */
    readonly theme?: IconTheme;
}

type DefaultProps = PickDefaultProps<IconProps, 'color' | 'padded'>;

export const defaultProps: DefaultProps = {
    color: 'default',
    padded: 'none'
};

class Icon extends PureComponent<IconProps> {
    public static readonly defaultProps = defaultProps;

    public render(): JSX.Element |Â null {
        const {padded, theme, label, icon, className, color, ...rest} = this.props;
        const iconClassName = icon;
        const classNames = mergeClassNames(
            theme!.icon,
            iconClassName,
            className,
            {
                [theme!['icon--paddedLeft']]: padded === 'left',
                [theme!['icon--paddedRight']]: padded === 'right',
                [theme!['icon--color-warn']]: color === 'warn',
                [theme!['icon--color-error']]: color === 'error',
                [theme!['icon--color-primaryBlue']]: color === 'primaryBlue',
            }
        );

        return <FontAwesomeIcon icon={this.getIconProp(icon) || 'question'} aria-label={label} className={classNames} {...rest} />;
    }

    private readonly getIconProp = (icon: string): IconProp => {
        const mappedIcon = mapper(icon);
        const iconArray = mappedIcon.split(' ');
        if (iconArray.length > 1) {
            const prefix = iconArray[0];
            const processedIcon = iconArray[1].startsWith('fa-') ? iconArray[1].substr(3) : iconArray[1];
            return [prefix as IconPrefix, processedIcon as IconName];
        } else {
            const prefix: IconPrefix = 'fas';
            return [prefix, mappedIcon as IconName];
        }
    }
}

export default Icon;
