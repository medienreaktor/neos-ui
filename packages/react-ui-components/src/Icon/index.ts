import {config, IconPrefix, library} from '@fortawesome/fontawesome-svg-core';
import '@fortawesome/fontawesome-svg-core/styles.css';
import {fab} from '@fortawesome/free-brands-svg-icons';
import {far} from '@fortawesome/free-regular-svg-icons';
import {fas} from '@fortawesome/free-solid-svg-icons';
import {themr} from '@friendsofreactjs/react-css-themr';

import identifiers from '../identifiers';
import Icon from './icon';

import style from './style.css';

// tslint:disable:no-object-mutation
config.autoAddCss = false;
config.familyPrefix = 'neos-fa' as IconPrefix;
config.replacementClass = 'neos-svg-inline--fa';
// tslint:enable:no-object-mutation

library.add(fab, fas, far);

export default themr(identifiers.icon, style)(Icon);
