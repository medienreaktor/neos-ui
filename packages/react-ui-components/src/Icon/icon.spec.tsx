import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';

import Icon, {defaultProps, IconProps} from './icon';

describe('<Icon/>', () => {
    const props: IconProps = {
        ...defaultProps,
        icon: 'fooIconClassName',
        theme: {
            'icon': 'iconClassName',
            'icon--big': 'bigClassName',
            'icon--color-error': 'color-errorClassName',
            'icon--color-primaryBlue': 'color-primaryBlueClassName',
            'icon--color-warn': 'color-warnClassName',
            'icon--paddedLeft': 'paddedLeftClassName',
            'icon--paddedRight': 'paddedRightClassName',
            'icon--small': 'smallClassName',
            'icon--spin': 'spinClassName',
            'icon--tiny': 'tinyClassName',
        }
    };

    it('should render correctly.', () => {
        const wrapper = shallow(<Icon {...props}/>);

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should allow the propagation of "className" with the "className" prop.', () => {
        const wrapper = shallow(<Icon {...props} className="fooClassName"/>);

        expect(wrapper.prop('className')).toContain('fooClassName');
    });

    it('should call the "fontAwesome.getClassName" api method and render the returned className.', () => {
        const wrapper = shallow(<Icon {...props}/>);

        expect(wrapper.hasClass('fooIconClassName')).toBeTruthy();
    });

    it('should allow the propagation of custom "icon" with the "icon" prop.', () => {
        const wrapper = shallow(<Icon {...props} icon="bazIconClassName"/>);

        expect(wrapper.hasClass('bazIconClassName')).toBeTruthy();
    });
});

