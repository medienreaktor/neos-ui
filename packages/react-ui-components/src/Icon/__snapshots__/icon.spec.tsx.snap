// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Icon/> should render correctly. 1`] = `
<FontAwesomeIcon
  border={false}
  className="iconClassName fooIconClassName"
  fixedWidth={false}
  flip={null}
  icon={
    Array [
      "fas",
      "fooIconClassName",
    ]
  }
  inverse={false}
  listItem={false}
  mask={null}
  pull={null}
  pulse={false}
  rotation={null}
  size={null}
  spin={false}
  symbol={false}
  title=""
  transform={null}
/>
`;
